osu.Game/Beatmaps/BeatmapImporter.cs

```diff
@@ -34,7 +34,7 @@ public class BeatmapImporter : RealmArchiveModelImporter<BeatmapSetInfo>
 
         protected override string[] HashableFileTypes => new[] { ".osu" };
 
-        public Action<BeatmapSetInfo>? ProcessBeatmap { private get; set; }
+        public Action<BeatmapSetInfo, bool>? ProcessBeatmap { private get; set; }
```

Comment:
I'm not sure about this. Should I pass the full `BeatmapUpdater` through or make this a delegate class or is this okay?

Reply:
I would say a delegate class would be preferable, at first glance I thought this `bool` was providing success results.

Reply:
Or a named tuple? `Action<(BeatmapSetInfo set, bool isBatch)>`

I tend to prefer not using delegate definitions because they don't handle great with intellisense/r# operations.

Reply:
While that means doubling up the parentheses when invoking the action, it does seem more organised comparing with an action or a delegate, I think I can get behind it. @smoogipoo what's your thoughts on this, here's a diff for reference.

```diff
diff --git a/osu.Game/Beatmaps/BeatmapImporter.cs b/osu.Game/Beatmaps/BeatmapImporter.cs
index df8b18313c..5f3fbcf27f 100644
--- a/osu.Game/Beatmaps/BeatmapImporter.cs
+++ b/osu.Game/Beatmaps/BeatmapImporter.cs
@@ -34,7 +34,7 @@ public class BeatmapImporter : RealmArchiveModelImporter<BeatmapSetInfo>
 
         protected override string[] HashableFileTypes => new[] { ".osu" };
 
-        public Action<BeatmapSetInfo, bool>? ProcessBeatmap { private get; set; }
+        public Action<(BeatmapSetInfo set, bool isBatch)>? ProcessBeatmap { private get; set; }
 
         public BeatmapImporter(Storage storage, RealmAccess realm)
             : base(storage, realm)
@@ -171,8 +171,7 @@ protected override void PreImport(BeatmapSetInfo beatmapSet, Realm realm)
         protected override void PostImport(BeatmapSetInfo model, Realm realm, bool batchImport)
         {
             base.PostImport(model, realm, batchImport);
-
-            ProcessBeatmap?.Invoke(model, batchImport);
+            ProcessBeatmap?.Invoke((model, batchImport));
         }
 
         private void validateOnlineIds(BeatmapSetInfo beatmapSet, Realm realm)
diff --git a/osu.Game/Beatmaps/BeatmapManager.cs b/osu.Game/Beatmaps/BeatmapManager.cs
index abf3d43d94..2782457ce7 100644
--- a/osu.Game/Beatmaps/BeatmapManager.cs
+++ b/osu.Game/Beatmaps/BeatmapManager.cs
@@ -42,7 +42,7 @@ public class BeatmapManager : ModelManager<BeatmapSetInfo>, IModelImporter<Beatm
 
         private readonly WorkingBeatmapCache workingBeatmapCache;
 
-        public Action<BeatmapSetInfo, bool>? ProcessBeatmap { private get; set; }
+        public Action<(BeatmapSetInfo set, bool isBatch)>? ProcessBeatmap { private get; set; }
 
         public BeatmapManager(Storage storage, RealmAccess realm, RulesetStore rulesets, IAPIProvider? api, AudioManager audioManager, IResourceStore<byte[]> gameResources, GameHost? host = null,
                               WorkingBeatmap? defaultBeatmap = null, BeatmapDifficultyCache? difficultyCache = null, bool performOnlineLookups = false)
@@ -62,7 +62,7 @@ public class BeatmapManager : ModelManager<BeatmapSetInfo>, IModelImporter<Beatm
             BeatmapTrackStore = audioManager.GetTrackStore(userResources);
 
             beatmapImporter = CreateBeatmapImporter(storage, realm);
-            beatmapImporter.ProcessBeatmap = (obj, isBatch) => ProcessBeatmap?.Invoke(obj, isBatch);
+            beatmapImporter.ProcessBeatmap = args => ProcessBeatmap?.Invoke(args);
             beatmapImporter.PostNotification = obj => PostNotification?.Invoke(obj);
 
             workingBeatmapCache = CreateWorkingBeatmapCache(audioManager, gameResources, userResources, defaultBeatmap, host);
@@ -323,7 +323,7 @@ public virtual void Save(BeatmapInfo beatmapInfo, IBeatmap beatmapContent, ISkin
 
                     setInfo.CopyChangesToRealm(liveBeatmapSet);
 
-                    ProcessBeatmap?.Invoke(liveBeatmapSet, false);
+                    ProcessBeatmap?.Invoke((liveBeatmapSet, false));
                 });
             }
 
diff --git a/osu.Game/OsuGameBase.cs b/osu.Game/OsuGameBase.cs
index 41eeece76d..3f5842bcfe 100644
--- a/osu.Game/OsuGameBase.cs
+++ b/osu.Game/OsuGameBase.cs
@@ -287,7 +287,7 @@ private void load(ReadableKeyCombinationProvider keyCombinationProvider)
 
             AddInternal(new BeatmapOnlineChangeIngest(beatmapUpdater, realm, metadataClient));
 
-            BeatmapManager.ProcessBeatmap = (set, isBatch) => beatmapUpdater.Process(set, preferOnlineFetch: !isBatch);
+            BeatmapManager.ProcessBeatmap = args => beatmapUpdater.Process(args.set, !args.isBatch);
 
             dependencies.Cache(userCache = new UserLookupCache());
             AddInternal(userCache);
```

Reply:
A tuple, a `ProcessBeatmapCallback` delegate, or a record struct (`ProcessBeatmapEvent`?) would all work, in order of personal preference.

Reply:
As in the first is the most preferable

Reply:
That's what I meant, yes.